### DISCUSSION TITLE
Monash datasets used for training Chronos

### DISCUSSION THREAD

**Discussion by tomar-s** [amazon-science/chronos-forecasting#167] [Q&A]
In the in-domain.yaml which says "# Backtest configs for the 15 "in-domain" datasets. The training portion of these datasets was part of the training corpus for Chronos models." 

I just wanted to confirm that you used the training part of ( _, test_template = split(gts_dataset, offset=offset)) load_and_split_dataset() function used in the evaluate.py as training dataset.

**Comment by abdulfatir** [comment]
Yes, the `_` which is skipped was used for training. As described in the paper, we first constructed augmentations from this training portion and then used the augmentations for training. The exact training corpus is available here: https://huggingface.co/datasets/autogluon/chronos_datasets#chronos-training-corpus-with-tsmixup--kernelsynth

**Comment by tomar-s** [comment]
Thank you for the confirmation. I was trying to use the training part(the skipped one) in zero shot datasets to finetune the models and check the performance on test portion of the same dataset after finetuning with its training part.
One more doubt there, why do you load the train split of the hf dataset in this function: load_and_split_dataset() for evaluation to get the test template

ds = datasets.load_dataset(
hf_repo, dataset_name, split="train", trust_remote_code=trust_remote_code
)

**Reply by abdulfatir** [reply]
As mentioned in the HF dataset repo:

> **NOTE**: The train split of all datasets contains the full time series and has no relation to the train/test split used in the Chronos paper.

**Reply by tomar-s** [reply]
Thanks for clarifying. Apologies for bothering you with so many questions. But when we want to fine tune on a hf dataset for example, monash_tourism_monthly which has a prediction length of 24 and offset -24. Should we be passing the same in the yaml file(chronos-t5-tiny.yaml) below:

Changing only these 3 values:

- Prediction length : 24 instead of 64
- context_length :  24 instead of 512
- path to the arrow file containing the training part of hf dataset (the _ which is skipped was used for training)

Below is the config I am looking to change: 
training_data_paths:
**"/home/ubuntu/tsmixup-data.arrow"**
**"/home/ubuntu/kernelsynth-data.arrow"** 
probability:
- 0.9
- 0.1
**context_length: 512
prediction_length: 64**
min_past: 60
max_steps: 200_000
save_steps: 100_000
log_steps: 500
per_device_train_batch_size: 32
learning_rate: 0.001
optim: adamw_torch_fused
num_samples: 20
shuffle_buffer_length: 100_000
gradient_accumulation_steps: 1
model_id: google/t5-efficient-tiny
model_type: seq2seq
random_init: true
tie_embeddings: true
output_dir: ./output/
tf32: true
torch_compile: true
tokenizer_class: "MeanScaleUniformBins"
tokenizer_kwargs:
  low_limit: -15.0
  high_limit: 15.0
n_tokens: 4096
lr_scheduler_type: linear
warmup_ratio: 0.0
dataloader_num_workers: 1
max_missing_prop: 0.9
use_eos_token: true

**Reply by abdulfatir** [reply]
@tomar-s First things first, are you able to convert the HF dataset into the arrow format? Can you tell me what steps you followed for that?

**Reply by abdulfatir** [reply]
Just to clarify on offset and prediction length. Offset is the index where you slice. The series is sliced into to portions at this index. The left portion is typically your training series. Now, the `prediction_length` converts the first `prediction_length` steps into the "label" which is used for evaluation. Note that in this case the entire right slice is used for evaluation as the prediction length is also 24.

**Reply by tomar-s** [reply]
This is the script I used to create training arrow files for each zero shot hf dataset.

```
from pathlib import Path
from typing import List, Union
import pandas as pd
import numpy as np
from gluonts.dataset.arrow import ArrowWriter
from gluonts.dataset.split import split
import datasets
import pyarrow.feather as feather
offset_alias_to_period_alias = {
    "WEEKDAY": "D",
    "EOM": "M",
    "BME": "M",
    "SME": "M",
    "BQS": "Q",
    "QS": "Q",
    "BQE": "Q",
    "BQE-DEC": "Q",
    "BQE-JAN": "Q",
    "BQE-FEB": "Q",
    "BQE-MAR": "Q",
    "BQE-APR": "Q",
    "BQE-MAY": "Q",
    "BQE-JUN": "Q",
    "BQE-JUL": "Q",
    "BQE-AUG": "Q",
    "BQE-SEP": "Q",
    "BQE-OCT": "Q",
    "BQE-NOV": "Q",
    "MS": "M",
    "D": "D",
    "B": "B",
    "min": "min",
    "s": "s",
    "ms": "ms",
    "us": "us",
    "ns": "ns",
    "h": "h",
    "QE": "Q",
    "QE-DEC": "Q-DEC",
    "QE-JAN": "Q-JAN",
    "QE-FEB": "Q-FEB",
    "QE-MAR": "Q-MAR",
    "QE-APR": "Q-APR",
    "QE-MAY": "Q-MAY",
    "QE-JUN": "Q-JUN",
    "QE-JUL": "Q-JUL",
    "QE-AUG": "Q-AUG",
    "QE-SEP": "Q-SEP",
    "QE-OCT": "Q-OCT",
    "QE-NOV": "Q-NOV",
    "YE": "Y",
    "YE-DEC": "Y-DEC",
    "YE-JAN": "Y-JAN",
    "YE-FEB": "Y-FEB",
    "YE-MAR": "Y-MAR",
    "YE-APR": "Y-APR",
    "YE-MAY": "Y-MAY",
    "YE-JUN": "Y-JUN",
    "YE-JUL": "Y-JUL",
    "YE-AUG": "Y-AUG",
    "YE-SEP": "Y-SEP",
    "YE-OCT": "Y-OCT",
    "YE-NOV": "Y-NOV",
    "W": "W",
    "ME": "M",
    "Y": "Y",
    "BYE": "Y",
    "BYE-DEC": "Y",
    "BYE-JAN": "Y",
    "BYE-FEB": "Y",
    "BYE-MAR": "Y",
    "BYE-APR": "Y",
    "BYE-MAY": "Y",
    "BYE-JUN": "Y",
    "BYE-JUL": "Y",
    "BYE-AUG": "Y",
    "BYE-SEP": "Y",
    "BYE-OCT": "Y",
    "BYE-NOV": "Y",
    "YS": "Y",
    "BYS": "Y",
    "QS-JAN": "Q",
    "QS-FEB": "Q",
    "QS-MAR": "Q",
    "QS-APR": "Q",
    "QS-MAY": "Q",
    "QS-JUN": "Q",
    "QS-JUL": "Q",
    "QS-AUG": "Q",
    "QS-SEP": "Q",
    "QS-OCT": "Q",
    "QS-NOV": "Q",
    "QS-DEC": "Q",
    "BQS-JAN": "Q",
    "BQS-FEB": "Q",
    "BQS-MAR": "Q",
    "BQS-APR": "Q",
    "BQS-MAY": "Q",
    "BQS-JUN": "Q",
    "BQS-JUL": "Q",
    "BQS-AUG": "Q",
    "BQS-SEP": "Q",
    "BQS-OCT": "Q",
    "BQS-NOV": "Q",
    "BQS-DEC": "Q",
    "YS-JAN": "Y",
    "YS-FEB": "Y",
    "YS-MAR": "Y",
    "YS-APR": "Y",
    "YS-MAY": "Y",
    "YS-JUN": "Y",
    "YS-JUL": "Y",
    "YS-AUG": "Y",
    "YS-SEP": "Y",
    "YS-OCT": "Y",
    "YS-NOV": "Y",
    "YS-DEC": "Y",
    "BYS-JAN": "Y",
    "BYS-FEB": "Y",
    "BYS-MAR": "Y",
    "BYS-APR": "Y",
    "BYS-MAY": "Y",
    "BYS-JUN": "Y",
    "BYS-JUL": "Y",
    "BYS-AUG": "Y",
    "BYS-SEP": "Y",
    "BYS-OCT": "Y",
    "BYS-NOV": "Y",
    "BYS-DEC": "Y",
    "Y-JAN": "Y-JAN",
    "Y-FEB": "Y-FEB",
    "Y-MAR": "Y-MAR",
    "Y-APR": "Y-APR",
    "Y-MAY": "Y-MAY",
    "Y-JUN": "Y-JUN",
    "Y-JUL": "Y-JUL",
    "Y-AUG": "Y-AUG",
    "Y-SEP": "Y-SEP",
    "Y-OCT": "Y-OCT",
    "Y-NOV": "Y-NOV",
    "Y-DEC": "Y-DEC",
    "Q-JAN": "Q-JAN",
    "Q-FEB": "Q-FEB",
    "Q-MAR": "Q-MAR",
    "Q-APR": "Q-APR",
    "Q-MAY": "Q-MAY",
    "Q-JUN": "Q-JUN",
    "Q-JUL": "Q-JUL",
    "Q-AUG": "Q-AUG",
    "Q-SEP": "Q-SEP",
    "Q-OCT": "Q-OCT",
    "Q-NOV": "Q-NOV",
    "Q-DEC": "Q-DEC",
    "W-MON": "W-MON",
    "W-TUE": "W-TUE",
    "W-WED": "W-WED",
    "W-THU": "W-THU",
    "W-FRI": "W-FRI",
    "W-SAT": "W-SAT",
    "W-SUN": "W-SUN",
}



def to_gluonts_univariate(hf_dataset: datasets.Dataset):
    series_fields = [
        col
        for col in hf_dataset.features
        if isinstance(hf_dataset.features[col], datasets.Sequence)
    ]
    series_fields.remove("timestamp")
    dataset_length = hf_dataset.info.splits["train"].num_examples * len(series_fields)
    dataset_freq = pd.infer_freq(hf_dataset[0]["timestamp"])
    dataset_freq = offset_alias_to_period_alias.get(dataset_freq, dataset_freq)

    gts_dataset = []
    for hf_entry in hf_dataset:
        for field in series_fields:
            gts_dataset.append(
                {
                    "start": pd.Period(
                        hf_entry["timestamp"][0],
                        freq=dataset_freq,
                    ),
                    "target": hf_entry[field],
                }
            )
    assert len(gts_dataset) == dataset_length

    return gts_dataset


def load_and_split_dataset(backtest_config: dict):

    hf_repo = backtest_config["hf_repo"]
    dataset_name = backtest_config["name"]
    offset = backtest_config["offset"]
    prediction_length = backtest_config["prediction_length"]
    num_rolls = backtest_config["num_rolls"]

    # This is needed because the datasets in autogluon/chronos_datasets_extra cannot
    # be distribued due to license restrictions and must be generated on the fly
    trust_remote_code = True if hf_repo == "autogluon/chronos_datasets_extra" else False

    ds = datasets.load_dataset(
        hf_repo, dataset_name, split="train", trust_remote_code=trust_remote_code
    )
    ds.set_format("numpy")

    gts_dataset = to_gluonts_univariate(ds)

    # Split dataset for evaluation
    _, test_template = split(gts_dataset, offset=offset)
    test_data = test_template.generate_instances(prediction_length, windows=num_rolls)

    return test_data


def convert_to_arrow(
      path: Union[str, Path],
      time_series: Union[List[np.ndarray], np.ndarray],
      compression: str = "lz4",
  ):
      """
      Store a given set of series into Arrow format at the specified path.

      Input data can be either a list of 1D numpy arrays, or a single 2D
      numpy array of shape (num_series, time_length).
      """
      assert isinstance(time_series, list) or (
          isinstance(time_series, np.ndarray) and
          time_series.ndim == 2
      )

      # Set an arbitrary start time
      start = np.datetime64("2000-01-01 00:00", "s")

      dataset = [
          {"start": start, "target": ts} for ts in time_series
      ]

      ArrowWriter(compression=compression).write_to_file(
          dataset,
          path=path,
      )


if __name__ == "__main__":
   

      ds = datasets.load_dataset(
          "autogluon/chronos_datasets", "monash_tourism_yearly", split="train"
      )
      ds.set_format("numpy")

      gts_dataset = to_gluonts_univariate(ds)

      # Split dataset for evaluation
      train_template, test_template = split(gts_dataset, offset=-4)

      ArrowWriter(compression="lz4").write_to_file(train_template, path="/Users/shivanitomar/Documents/Implementations/FM_TUNE/fm-tune/chronos_datasets/monash_tourism_yearly.arrow")

      # Load the Feather file
      table = feather.read_table('./sample.arrow')
      # Convert to Pandas DataFrame
      df = table.to_pandas()
```

**Reply by tomar-s** [reply]
What would be the correct approach to use this particular dataset to both finetune and evaluate on?
I am confused between the offset, context_length and prediction_length config passed in train and evaluate.py.
Where exactly am I going wrong with these parameters?

**Reply by tomar-s** [reply]
I've tried to go through the paper again to check if I am missing something. But I couldn't find any reference on how to configure the yaml files when finetuning the model on a specific task (in my case the zero shot monash datasets). 
And likewise how to pass offset and prediction length when evaluating on the same dataset you fine-tuned your model on. You also mention in the paper the positive impact of longer context length on the evalutaion metric MASE.  
I am particularly looking for information how to set these parameters right when finetuning and evaluating on the same hf zero shot dataset:
  -min_past
  -context length
  -offset
  -prediction_length
  -num_rolls
 I looked through the repo to find a more detailed description on these parameters but couldn't find it. Can you please provide your guidance on how to correctly pass this configuration.

**Reply by abdulfatir** [reply]
@tomar-s your script looks okay to me. You can use the generated arrow file for fine-tuning. For fine-tuning on a specific dataset, you should change the `prediction_length` during fine-tuning to the desired prediction length. For example, if you want to evaluate a dataset with a prediction horizon of 24, you might want to fine-tune the model with that prediction length. You can either leave the context length to 512 or reduce it if your dataset has shorter series. The `min_past` parameters sets the expected minimum context available for the model. This means that any time series which is shorter than `min_past + prediction_length` will be skipped during training. If you don't want to skip anything during fine-tuning, set `min_past` to 0. I hope this helps. Lmk if you face any other problems.

**Reply by tomar-s** [reply]
Thanks for your response. I will get back if I face further issues. In the meantime, do you have any documentation describing the parameters in both train and evaluate configs just how you explained them above. For examples: num_rolls though it is set to 1 for all evaluation datasets but how does changing this value affect the particular dataset.

**Comment by tomar-s** [comment]
I see that while training for in-domain monash datasets you set the context length to 512 (for all datasets) and prediction length to 64 which is  greater than the prediction lengths of all datasets considered for evaluation. As you mentioned in your previous response, I can keep the context length as 512 or smaller based on the size of the time series in the specific dataset picked for fine-tuning. My question is when I compare my results(MASE) of finetuned tiny/mini model with its zero shot performance on that same dataset will this context length not affect the results?

For example like the dataset NN5 weekly has series length of only 113 so we need to set the context length <512 to lets say 48 or 36. and prediction length to 8 for this dataset. Then I will evaluate the fine-tuned model on this dataset setting offset and prediction length as -8 and 8 respectively. The MASE will be compared to the zero shot performance on this dataset which acc to the table in paper is 0.927. Is this comparison fair? The shorter context length I've set while finetuning will not have any impact?
How do you handle context lengths for short time series length datasets present in the in-domain benchmark like Electricity weekly or M4 monthly?

**Reply by abdulfatir** [reply]
In the simple fine-tuning experiment we did in the paper, we used the same context length of 512 for all datasets. However, depending on your use case this may not be an ideal choice and you may want to change the context length for your downstream dataset. Note that during inference the model will only use the available context.

**Comment by tomar-s** [comment]
I don't have any specific dataset for fine-tuning so I am using the zero-shot datasets from your paper where the series length is atleast 512 to keep the context length=512 as set in the pre-trained model. I am specifically using traffic, weather, fred-md, australian electricity to name a few.
